src/resolve/adapters/baseSourceAdapter.ts

```diff
@@ -146,6 +146,24 @@ export abstract class BaseSourceAdapter implements SourceAdapter {
     }
   }
 
+  private calculateName(rootMetadata: MetadataXml): string {
+    if (!this.type.folderType) {
+      return rootMetadata.fullName;
+    }
+    const grandparentType = this.registry.getTypeByName(this.type.folderType);
+    if (grandparentType.folderType && grandparentType.folderType !== this.type.id) {
+      const splits = rootMetadata.path.split(sep);
+      return `${splits[splits.indexOf(grandparentType.directoryName) + 1]}.${
+        rootMetadata.fullName
+      }`;
+    }
+    if (grandparentType.folderType === this.type.id) {
+      return rootMetadata.fullName;
+    }
+
+    return `${parentName(rootMetadata.path)}/${rootMetadata.fullName}`;
+  }
```

Comment:
I think this could use some debug or at least a comment for each return providing an example.  Most of this logic is for folder types so maybe it would improve readability to do something like below this instead?

In `getComponent()` just before creating a new SourceComponent:
```javascript
let fullName = rootMetadata.fullName;
let parentType;
if (this.type.folderType) {
  parentType = this.registry.getTypeByName(this.type.folderType);
  fullName = getFolderFullName(rootMetadata,  parentType);
}
```

Then in getFolderFullName it would be mostly the same as calculateName just only about folders
```javascript
private getFolderFullName(rootMetadata, parentType) { ... }
```

Reply:
I added some explanatory comments with examples.

That's an elegant solution for the current situation.  I think we'll  have a few more flavors coming (sharing*Rule) where the names also have wrinkles.  

I was wanting one place to keep all that stuff.

Reply:
would it be ok to add our normal logger stuff in here?  What if there were like 1000 components?  BSA doesn't have a way to know the size...it would need to be told that somewhere.

Reply:
Maybe debug isn't a good option then.

Reply:
it's hard to do in SDR...you'd be passing it down multiple levels to get to some of stream transformer stuff.